core/src/main/java/com/sequenceiq/cloudbreak/service/RotateSaltPasswordService.java

```diff
@@ -86,19 +97,62 @@ private boolean isChangeSaltuserPasswordSupported(StackDto stack) {
+        RotateSaltPasswordType rotateSaltPasswordType = getRotateSaltPasswordType(stack);
+        LOGGER.info("Triggering rotate salt password for stack {} with type {}", stack.getId(), rotateSaltPasswordType);
+        return flowManager.triggerRotateSaltPassword(stack.getId(), reason, rotateSaltPasswordType);
+    }
+
+    private RotateSaltPasswordType getRotateSaltPasswordType(StackDto stack) {
+        return isChangeSaltuserPasswordSupported(stack) ? RotateSaltPasswordType.SALT_BOOTSTRAP_ENDPOINT : RotateSaltPasswordType.FALLBACK;
     }
 
     public void rotateSaltPassword(StackDto stack) throws CloudbreakOrchestratorException {
         validateRotateSaltPassword(stack);
-        SecurityConfig securityConfig = securityConfigService.getOneByStackId(stack.getId());
+        SecurityConfig securityConfig = stack.getSecurityConfig();
         String oldPassword = securityConfig.getSaltSecurityConfig().getSaltPassword();
         String newPassword = PasswordUtil.generatePassword();
         List<GatewayConfig> allGatewayConfig = gatewayConfigService.getAllGatewayConfigs(stack);
         hostOrchestrator.changePassword(allGatewayConfig, newPassword, oldPassword);
         securityConfigService.changeSaltPassword(securityConfig, newPassword);
     }
 
+    public void rotateSaltPasswordFallback(StackDto stack) throws CloudbreakOrchestratorFailedException {
+        List<GatewayConfig> allGatewayConfig = gatewayConfigService.getAllGatewayConfigs(stack);
+        tryRemoveSaltuserFromGateways(stack, allGatewayConfig);
+
+        String newPassword = PasswordUtil.generatePassword();
+        SecurityConfig securityConfig = stack.getSecurityConfig();
+        securityConfig.getSaltSecurityConfig().setSaltPassword(newPassword);
+        try {
+            InMemoryStateStore.putStack(stack.getId(), PollGroup.POLLABLE);
+            clusterBootstrapper.reBootstrapOnHost(stack);
+            securityConfigService.changeSaltPassword(securityConfig, newPassword);
+        } catch (CloudbreakException e) {
+            Set<String> gatewayConfigAddresses = allGatewayConfig.stream()
+                    .map(GatewayConfig::getPrivateAddress)
+                    .collect(Collectors.toSet());
+            LOGGER.warn("Failed to re-bootstrap gateway nodes after saltuser password delete", e);
+            String message = String.format("Failed to re-bootstrap gateway nodes after saltuser password delete. " +
+                            "Please check the salt-bootstrap service status on node(s) %s. " +
+                            "If the saltuser password was changed manually, " +
+                            "please remove the user manually with the command '%s' on node(s) %s and retry the operation.",
+                    gatewayConfigAddresses, SALTUSER_DELETE_COMMAND, gatewayConfigAddresses);
+            throw new CloudbreakOrchestratorFailedException(message, e);
+        } finally {
+            InMemoryStateStore.deleteStack(stack.getId());
```

Comment:
shouldn't this be handled with a status change also?

Reply:
It is handled here: https://github.com/hortonworks/cloudbreak/blob/814f987be9fdf6b2968e14cead38f6718e462a42/core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/cluster/salt/rotatepassword/RotateSaltPasswordActions.java#L73